docs/design/hierarchical-queue-on-capacity-plugin.md

```diff
@@ -0,0 +1,133 @@
+  	parents []api.QueueID
+
+  	deserved  *api.Resource
+  	allocated *api.Resource
+  	request   *api.Resource
+  	// elastic represents the sum of job's elastic resource, job's elastic = job.allocated - job.minAvailable
+  	elastic *api.Resource
+  	// inqueue represents the resource request of the inqueue job
+  	inqueue    *api.Resource
+  	capability *api.Resource
+  	// realCapability represents the resource limit of the queue, LessEqual capability
+  	realCapability *api.Resource
+  	guarantee      *api.Resource
+  	// children represents the child queues of the queue
+  	children map[api.QueueID]*queueAttr
+  }
+  ```
+
+- Extend the capacityPlugin struct in capacity plugin to include one new field:
+  - `rootQueue`: Represents the root queue. It is the parent of all queues.
+  
+  ```go
+  type capacityPlugin struct {
+    totalResource  *api.Resource
+    totalGuarantee *api.Resource
+    rootQueue api.QueueID
+    
+    queueOpts map[api.QueueID]*queueAttr
+    // Arguments given for the plugin
+    pluginArguments framework.Arguments
+  }
+  ```
+
+**Note:** When the scheduler has the capacity plugin enabled and `EnabledHierarchy` is set to true, a root queue with queueID `root` will be automatically created. In the `OpenSession` function, the `Deserved` and `Capability` fields of the root queue are updated each time based on `ssn.TotalResource`.
+
+### Key function modifications
+
+- `OnSessionOpen`: Check that the sum of the deserved/guarantee values of each queue's child queues does not exceed the deserved/guarantee of the parent queue, and that the capability of any child queue does not exceed that of the parent queue. 
+  - If the check reveals non-compliance with the aforementioned conditions, an error message should be printed, and the relevant functions for the session should return a disallowed status. 
+  - For example, in the case of `AddJobEnqueueableFn`, it should directly return `util.Reject`, preventing the job from being enqueued. No scheduling should be performed in situations that do not align with the previously described conditions. Users need to adjust the queues based on the printed error information until normal scheduling can be resumed.
+- `AddQueueOrderFn`: Prioritize leaf queues (queues without child queues) when considering queue ordering. For two leaf queues, the ordering will be determined by considering their parent queues layer by layer, starting from the root queue.
+- `AddAllocatableFn`: Ensure that only leaf queues can schedule jobs/podgroups.
+- `BuildVictimsPriorityQueue`: The current implementation of this function relies on `TaskOrderFn` and `JobOrderFn` for sorting victims. Two new functions, `VictimTaskOrderFn` and `VictimJobOrderFn`, will be introduced specifically for sorting victims in the context of hierarchical queues. 
+  
+  ```go
+  func (ssn *Session) BuildVictimsPriorityQueue(victims []*api.TaskInfo, preemptor *api.TaskInfo) *util.PriorityQueue {
```

Comment:
Seems that this logic should exist in `AddReclaimableFn`, right?

Reply:
AddReclaimableFn is used to select victims, while VictimTaskOrderFn and VictimJobOrderFn are used to sort the victims.